### PR TITLE
Fix ABI string docs in reference.md

### PR DESCRIPTION
**Pull request by strega-nil**

### CODE REVIEW

**Comment by ubsan** [comment]
I know it's not the best, but I'm not great at writing documentation. I just wanted to get a first attempt there.

File: src/doc/reference.md [hunk 1]
```diff
@@ -1631,14 +1631,41 @@ the Rust ABI and the foreign ABI.
...
+  itself
+
+There are also some platform-specific ABI strings:
+
+* `extern "cdecl"` -- The default for x86\_32 C code.
+* `extern "stdcall"` -- The default for the Win32 API on x86\_32.
+* `extern "win64"` -- The default for C code on x86\_64 Windows.
+* `extern "aapcs"` -- The default for ARM.
+* `extern "fastcall"`
+* `extern "vectorcall"`
```
Comment (eefriedman) [inline src/doc/reference.md]: `extern "fastcall"` corresponds to the Visual Studio's `__fastcall` and gcc's `__attribute__((fastcall))` is a.  `extern "vectorcall"` corresponds to the Visual Studio's `__vectorcall`.

**Comment by GuillaumeGomez** [comment]
No, that's a nice work. :+1: for me.

**Comment by steveklabnik** [comment]
This looks good to me, but I would like @rust-lang/compiler and @rust-lang/lang to make a call here about "all compilers are guaranteed to support"

File: src/doc/reference.md [hunk 2]
```diff
@@ -1631,14 +1631,43 @@ the Rust ABI and the foreign ABI.
...
+  `__fastcall` and GCC and clang's `__attribute__((fastcall))`
+* `extern "vectorcall"` -- The `vectorcall` ABI -- corresponds to MSVC's
+  `__vectorcall` and clang's `__attribute__((vectorcall))`
+
+Finally, there are some rustc-specific ABI strings:
+
+* `extern "rust-intrinsic"` -- The ABI of rustc intrinsics.
+* `extern "rust-call"` -- The ABI of the Fn::call trait functions.
+* `extern "platform-intrinsic"` -- Specific platform intrinsics -- like, for
+  example, `sqrt` -- have this ABI. You should never have to deal with it.
```
Comment (eddyb) [inline src/doc/reference.md]: `sqrt`? I thought that was a regular intrinsic. Also, a subset of the platform intrinsics is likely to be stabilized in some form, while the more general set of intrinsics will remain unstable indefinitely.

Comment (eddyb) [inline src/doc/reference.md]: The `simd` crate, for example, uses `platform-intrinsics`.

**Comment by ubsan** [comment]
@eddyb I wasn't sure. Is there... like... any documentation on this? or code I can read?

**Comment by eddyb** [comment]
@ubsan The existing implementation is governed by https://github.com/rust-lang/rfcs/blob/master/text/1199-simd-infrastructure.md AFAIK.

**Comment by alexcrichton** [comment]
ping, is this ready for merge?

File: src/doc/reference.md [hunk 3]
```diff
@@ -1631,14 +1631,43 @@ the Rust ABI and the foreign ABI.
...
+standard C ABI on the specific platform. Other ABIs may be specified using an
+`abi` string, as shown here:
 
 ```ignore
 // Interface to the Windows API
 extern "stdcall" { }
 ```
 
+There are three ABI strings which are cross-platform, and which all compilers
+are guaranteed to support:
```
Comment (nikomatsakis) [inline src/doc/reference.md]: Given the fact that there are no other rust compilers, it seems like we should just say "which are cross-platform". This wording suggests that other compilers exist.

Comment (ubsan) [inline src/doc/reference.md]: They do? They're not finished, but they exist.

Comment (eddyb) [inline src/doc/reference.md]: [mrustc](https://github.com/thepowersgang/mrustc) is the main one I know of.

File: src/doc/reference.md (hunk 2 above)
Comment (steveklabnik) [inline src/doc/reference.md]: @ubsan what's your response to @eddyb here?

File: src/doc/reference.md (hunk 3 above)
Comment (steveklabnik) [inline src/doc/reference.md]: @nikomatsakis and the rest of @rust-lang/lang , are we comfortable saying that all compilers are guaranteed to support this? It feels like a strange thing to "guarantee".

Comment (nrc) [inline src/doc/reference.md]: I think it is a requirement, not a guarantee - i.e., "which all compilers must support:"

**Pull request merged**